src/wasm/wasm-type.cpp

```diff
@@ -2203,6 +2291,258 @@ bool FiniteShapeEquator::eq(const Rtt& a, const Rtt& b) {
   return a.depth == b.depth && eq(a.heapType, b.heapType);
 }
 
+bool RecGroupStructure::operator==(const RecGroupStructure& other) const {
+  if (group == other.group) {
+    return true;
+  }
+  // The rec groups are equivalent if they are piecewise equivalent.
+  return std::equal(group.begin(),
+                    group.end(),
+                    other.group.begin(),
+                    other.group.end(),
+                    [&](const HeapType& a, const HeapType& b) {
+                      return topLevelEq(a, b, other.group);
+                    });
+}
+
+bool RecGroupStructure::topLevelEq(HeapType a,
+                                   HeapType b,
+                                   RecGroup other) const {
+  if (a == b) {
+    return true;
+  }
+  if (a.isBasic() || b.isBasic()) {
+    return false;
+  }
+  return eq(*getHeapTypeInfo(a), *getHeapTypeInfo(b), other);
+}
+
+bool RecGroupStructure::eq(Type a, Type b, RecGroup other) const {
+  if (a == b) {
+    return true;
+  }
+  if (a.isBasic() || b.isBasic()) {
+    return false;
+  }
+  return eq(*getTypeInfo(a), *getTypeInfo(b), other);
+}
+
+bool RecGroupStructure::eq(HeapType a, HeapType b, RecGroup other) const {
+  // Do not recurse into the structure of children `a` and `b`, but check
+  // whether their recursion groups and indices match. Since the wrapped group
```

Comment:
To make sure I follow, the "wrapped group" here refers to the group the RecGroupStructure itself wraps?

Reply:
Yes, correct.